pkg/api/validation/validation.go

```diff
@@ -1276,17 +1276,51 @@ func validateEndpointSubsets(subsets []api.EndpointSubset) errs.ValidationErrorL
 		if len(ss.Ports) == 0 {
 			ssErrs = append(ssErrs, errs.NewFieldRequired("ports"))
 		}
-		// TODO: validate each address and port
+		for addr := range ss.Addresses {
+			ssErrs = append(ssErrs, validateEndpointAddress(&ss.Addresses[addr])...)
+		}
+		for port := range ss.Ports {
+			ssErrs = append(ssErrs, validateEndpointPort(&ss.Ports[port], len(ss.Ports) > 1)...)
+		}
+
 		allErrs = append(allErrs, ssErrs.PrefixIndex(i)...)
 	}
 
 	return allErrs
 }
 
+func validateEndpointAddress(address *api.EndpointAddress) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+	if !util.IsValidIPv4(address.IP) {
+		allErrs = append(allErrs, errs.NewFieldInvalid("ip", address.IP, "is invalid IPv4 address"))
```

Comment:
I would drop the "is" in the error - it's not generally formatted as a sentence.